File: tree/diameter.py
class Node:
    def __init__(self,key):
        self.left=None
        self.right=None
        self.data=None

def H(root):
    if root is None:
        return;
    lh=H(root.left)
    rh=H(root.right)
    if lh>rh:
        return lh
    else:
        return rh

def D(root):
    if root is None:
        return;
    lh=H(root.left)
    rh=H(root.right)
    ld=D(root.left)
    rd=D(root.right)
    return max(lh+rh+1,max(ld,rd))
root=Node(1)
root.right=Node(2)
root.left=Node(3)
root.left.left=Node(4)
root.left.right=Node(5)
root.right.left=Node(6)
root.right.right=Node(7)
print(D(root))

File: tree/diameter.cpp
#include<stdio.h>
#include<stdlib.h>
struct node
{
    int data;
    struct node *left,*right;
};
struct node* newnode(int data)
{
    struct node* tr=(struct node*)malloc(sizeof(struct node));
    tr->data=data;
    tr->left=tr->right=NULL;
    return tr;
}
int maximum(int a,int b)
{
    return (a>b)?a:b;
}
int hight(struct node* tr)
{
    if(tr==NULL)
        return 0;
    return 1+maximum(hight(tr->left),hight(tr->right));
}
int diameter(struct node* tr)
{
    if(tr==NULL)
        return 0;
    int lhi=hight(tr->left);
    int rhi=hight(tr->right);
    int ldim=diameter(tr->left);
    int rdim=diameter(tr->right);
    return maximum(lhi+rhi+1,maximum(ldim,rdim));
}
int main()
{
    struct node* tr=newnode(12);
    tr->left=newnode(21);
    tr->right=newnode(23);
    tr->left->left=newnode(11);
    tr->left->right=newnode(10);
    tr->left->right->right=newnode(4);
    printf("%d",diameter(tr));
    return 0;
}
